Backpack/src/main/java/net/skyscanner/backpack/button/BpkButton.kt

```diff
@@ -0,0 +1,282 @@
+package net.skyscanner.backpack.button
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.content.res.ColorStateList
+import android.graphics.Color
+import android.graphics.drawable.Drawable
+import android.graphics.drawable.GradientDrawable
+import android.graphics.drawable.RippleDrawable
+import android.graphics.drawable.StateListDrawable
+import android.os.Build
+import android.support.annotation.ColorInt
+import android.support.annotation.ColorRes
+import android.support.v4.content.ContextCompat
+import android.support.v4.graphics.drawable.DrawableCompat
+import android.support.v7.widget.AppCompatButton
+import android.util.AttributeSet
+import net.skyscanner.backpack.R
+
+open class BpkButton @JvmOverloads constructor(
+  context: Context,
+  attrs: AttributeSet? = null,
+  defStyleAttr: Int = 0
+) : AppCompatButton(context, attrs, defStyleAttr) {
+
+  private val roundedButtonCorner by lazy(LazyThreadSafetyMode.NONE) { dpToPx(16, context).toFloat() }
+  private val strokeWidth by lazy(LazyThreadSafetyMode.NONE) { 2 }
+
+  var type: Type = Type.Primary
+    set(value) {
+      field = value
+      setup()
+    }
+
+  var size: Size = Size.Basic
+    set(value) {
+      field = value
+      setup()
+    }
+
+  var icon: Drawable? = null
+    set(value) {
+      if (value != null) {
+        field = value
+        setup()
+      }
+    }
+
+  init {
+    val attr = context.theme.obtainStyledAttributes(attrs, R.styleable.BpkButton, defStyleAttr, 0)
+    try {
+      type = Type.fromId(attr.getInt(R.styleable.BpkButton_buttonType, 0))
+      size = Size.fromId(attr.getInt(R.styleable.BpkButton_buttonSize, 0))
+      attr.getDrawable(R.styleable.BpkButton_buttonIcon)?.let { drawable ->
+        icon = DrawableCompat.wrap(drawable)
+      }
+    } finally {
+      attr.recycle()
+    }
+    setup()
+  }
+
+  enum class Type(internal var id: Int, @ColorRes internal var bgColor: Int, @ColorRes internal var textColor: Int, @ColorRes internal var strokeColor: Int) {
+    Primary(0, R.color.bpkGreen500, R.color.bpkWhite, android.R.color.transparent),
+    Secondary(1, R.color.bpkWhite, R.color.bpkBlue600, R.color.bpkGray100),
+    Featured(2, R.color.bpkPink500, R.color.bpkWhite, android.R.color.transparent),
+    Destructive(3, R.color.bpkWhite, R.color.bpkRed500, R.color.bpkGray100);
+
+    internal companion object {
+      internal fun fromId(id: Int): Type {
+        for (f in values()) {
+          if (f.id == id) return f
+        }
+        throw IllegalArgumentException()
+      }
+    }
+  }
+
+  enum class Size(internal var id: Int) {
+    Basic(0),
```

Comment:
on iOS we have 'Default' and 'Large'. Not sure which one to go with

Reply:
`default` as an attribute in XML is not allowed by Android, therefore I chose another word for this. I wrote this in the Gotchas section of the PR. 

![image](https://user-images.githubusercontent.com/935897/45831021-95752e80-bcfe-11e8-913f-a214d0e1a164.png)